[fix][doc] Rewrite the steps of docker-deploy (#17853)
apache/pulsar: site2/docs/deploy-docker.md

id: deploy-docker
title: Deploy a cluster on Docker
sidebar_label: "Docker"

> —

Reviewer: This is an obvious syntax error that fails the site build.

Please be careful when reviewing...Always previewing your changes helps.

Author: I made a suggestion to modify the description template. I think we should process this to ensure that rather than hope that reviewers and authors are always right.

Reviewer: No. The community runs by meritocracy so reviewers should be careful. It's not irreversible to make mistake like this and I'm not blaming you here. But send a notice that you should follow the process. Adding more description hints you to follow the process also but I believe few people really read the whole description.

Reviewer 2: Shouldn't there be a sanity check CI job for any documentation change in the repo if full pipeline is too expensive?

You can use two kinds of methods to deploy a Pulsar cluster on Docker.
The first uses Docker commands, while the second uses a `docker-compose.yaml` file.
## Deploy a cluster using Docker commands
To deploy a Pulsar cluster on Docker, you need to complete the following steps:

> Pull a Pulsar Docker image.

Reviewer 3: ```suggestion
1. Pull a Pulsar Docker image.
```
The three steps missed the numbering. Can you pls add them all?

title: Deploy a cluster on Docker
sidebar_label: "Docker"
---
## Deploy a cluster on Docker

> To deploy a Pulsar cluster on Docker, you need to complete the next steps:

Reviewer 3: ```suggestion
To deploy a Pulsar cluster on Docker, you need to complete the following steps:
```

> 1. pull the pulsar docker image

Reviewer 3: ```suggestion
1. Pull the Pulsar Docker image.
```

> 2. create the zookeeper, bookie, broker container by the image

Reviewer 3: ```suggestion
2. Create ZooKeeper, bookie, and broker containers.
```

> 3. create a network, and make the container connects to it.

Reviewer 3: ```suggestion
3. Create a network connected to the containers.
```

> 4. start the zookeeper, and then init the cluster metadata

Reviewer 3: ```suggestion
4. Start ZooKeeper and initialize cluster metadata.
```

> 5. start the broker and bookie

Reviewer 3: ```suggestion
5. Start brokers and bookies.
```

## Prepare
To run Pulsar on Docker, you need to create a container for each Pulsar component: ZooKeeper, BookKeeper and broker. You can pull the images of ZooKeeper and BookKeeper separately on Docker Hub, and pull a Pulsar image for the broker. You can also pull only one Pulsar image and create three containers with this image. This tutorial takes the second option as an example.

## Pull a Pulsar image
You can pull a Pulsar image from Docker Hub with the following command. If you want to use some connectors, you can use apachepulsar/pulsar-all there.

```java
docker pull apachepulsar/pulsar-all:latest
```
## Create containers

> * Create zookeeper container

Reviewer 3: ```suggestion
* Create a ZooKeeper container
```

> ```

Reviewer 3: Does it make sense to add a language highlight like `bash` after ```? If yes, it applies to all the following commands.

docker run -d --privileged=true -u=root -e PULSAR_PREFIX_metadataStoreUrl=zk:zookeeper:2181 -e PULSAR_PREFIX_cluster-name=cluster-a -e PULSAR_PREFIX_managedLedgerDefaultEnsembleSize=1 -e PULSAR_PREFIX_managedLedgerDefaultWriteQuorum=1 -e PULSAR_PREFIX_managedLedgerDefaultAckQuorum=1 --name zookeeper apachepulsar/pulsar-all:latest /bin/bash
```

> * Create broker container

Reviewer 3: ```suggestion
* Create a broker container
```

```
docker run -d --privileged=true -u=root -e PULSAR_PREFIX_metadataStoreUrl=zk:zookeeper:2181 -e PULSAR_PREFIX_cluster-name=cluster-a -e PULSAR_PREFIX_managedLedgerDefaultEnsembleSize=1 -e PULSAR_PREFIX_managedLedgerDefaultWriteQuorum=1 -e PULSAR_PREFIX_managedLedgerDefaultAckQuorum=1 --name broker apachepulsar/pulsar-all:latest /bin/bash
```

> * Create bookie container

Reviewer 3: ```suggestion
* Create a bookie container
```

```
docker run -d --privileged=true -u=root -e PULSAR_PREFIX_zkServers=zookeeper:2181 -e PULSAR_PREFIX_metadataServiceUri=metadata-store:zk:zookeeper:2181  --name bookie apachepulsar/pulsar-all:latest /bin/bash
```
## Create the network
To deploy a Pulsar cluster on Docker, you need to create a network and connect the containers of ZooKeeper, BookKeeper and broker to this network. The following command creates the network pulsar:
```
docker network create pulsar
```
Connect the containers of ZooKeeper, BookKeeper and broker to the pulsar network with the following commands.

```
docker network connect pulsar zookeeper
```
```
docker network connect pulsar bookkeeper
```
```
docker network connect pulsar broker
```

> ## start the zookeeper, and init the cluster metadata

Reviewer 3: ```suggestion
## Start ZooKeeper and initialize cluster metadata
```

> Start the zookeeper service in the docker container by the following commands:

Reviewer 3: ```suggestion
Start the ZooKeeper service in the Docker container by running the following commands.
```

> 1. Open the zookeeper container

Reviewer 3: ```suggestion
1. Open the ZooKeeper container.
```

```
docker exec -it zookeeper bash
```

> 2. Start the zookeeper service

Reviewer 3: ```suggestion
2. Start the ZooKeeper service.
```

```
bin/pulsar-daemon start zookeeper
```

> 3. init the cluster metadata

Reviewer 3: ```suggestion
3. Initialize the cluster metadata.
```

```
 bin/pulsar initialize-cluster-metadata \
          --cluster cluster-a \
          --zookeeper zookeeper:2181 \
          --configuration-store zookeeper:2181 \
          --web-service-url http://broker:8080 \
          --broker-service-url pulsar://broker:6650 \
```

> ## start the broker and bookie service

Reviewer 3: ```suggestion
## Start broker and bookie services
```

> Start the broker and bookie service in the docker container by the following commands:

Reviewer 3: ```suggestion
Start the broker and bookie services in the Docker container by running the following commands.
```

### Start broker service
1. Open the broker container
```
docker exec -it broker bash
```
2. Start the broker service
```
bin/pulsar-daemon start broker
```

### Start bookie service
1. Open the bookie container
```
docker exec -it bookie bash
```
2. Start the bookie service
```

> bin/pulsar-daemon start bookie

Reviewer 4: Can you add an example of using a local pulsar client to produce and consume by connecting to the docker pulsar service?

Author: This is a deployment instead of ` get start`, right?
But I think we can export the ports.
And you can comment on the google doc.
